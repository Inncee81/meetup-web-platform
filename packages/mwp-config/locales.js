module.exports = [
	'en-US',	
	'en-AU',
	'es',
	'es-ES',
	'de',
	'de-DE',
	'fr',
	'fr-FR',
	'it',
	'it-IT',
	'ja',
	'ja-JP',
	'ko',
	'ko-KR',
	'nl',	
	'nl-NL',
	'pt',
	'pt-BR',
	'pl',
	'pl-PL',
	'ru',
	'ru-RU',
	'th',
	'th-TH',
	'tr',
	'tr-TR',
];
