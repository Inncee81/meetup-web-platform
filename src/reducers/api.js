/**
 * The root level reducer for the app.
 * @module reducer
 **/

import { combineReducers } from 'redux';
import {
	CLICK_TRACK_ACTION,
	CLICK_TRACK_CLEAR_ACTION
} from '../actions/clickActionCreators';

export const DEFAULT_APP_STATE = { isFetching: false };

/**
 * The primary reducer for data provided by the API
 * `state.app` sub-tree
 *
 * @param {Object} state
 * @param {ReduxAction} action
 * @return {Object}
 */
export function app(state=DEFAULT_APP_STATE, action={}) {
	switch (action.type) {
	case 'API_REQUEST':
		if (action.meta.logout) {
			// clear app state during logout
			return { ...DEFAULT_APP_STATE, isFetching: true };
		}
		return { ...state, isFetching: true };
	case 'API_SUCCESS':  // fall though
	case 'CACHE_SUCCESS':  // fall through
	case 'API_ERROR': {
		// each of these actions provides an API response that should go into app
		// state - error responses will contain error info
		delete state.failure;  // if there are any values, the API is not failing
		const { ref, ...response } = action.payload.response;
		return { ...state, [ref]: response };
	}
	case 'API_FAILURE':
		state.failure = action.payload;
		// fall through - fetch is complete
	case 'API_COMPLETE':
		return { ...state, isFetching: false };
	default:
		return state;
	}
}

export const DEFAULT_CLICK_TRACK = { history: [] };
/**
 * @param {Object} data extensible object to store click data {
 *   history: array
 * }
 * @param {Object} action the dispatched action
 * @return {Object} new state
 */
export function clickTracking(state=DEFAULT_CLICK_TRACK, action) {
	if (action.type === CLICK_TRACK_ACTION) {
		const history = [
			...state.history,
			action.payload,
		];
		return {
			...state,
			history,
		};
	}
	if (action.type === CLICK_TRACK_CLEAR_ACTION) {
		return DEFAULT_CLICK_TRACK;
	}
	return state;
}

export function config(state={}, action) {
	switch(action.type) {
	case 'CONFIGURE_API_URL':
		return { ...state, apiUrl: action.payload };
	case 'CONFIGURE_BASE_URL':
		return { ...state, baseUrl: action.payload };
	default:
		return state;
	}
}

/**
 * This reducer manages a list of boolean flags that indicate the 'ready to
 * render' state of the application. It is used exclusively by the server,
 * which triggers actions when initializing a response that should eventually
 * make all flags 'true'
 *
 * The server can then read these flags from state and render when ready
 */
export function preRenderChecklist([apiDataLoaded] = [false], action) {
	return [
		apiDataLoaded || action.type === 'API_COMPLETE',
	];
}

const platformReducers = {
	app,
	clickTracking,
	config,
	preRenderChecklist,
};

/**
 * A function that builds a reducer combining platform-standard reducers and
 * app-specific reducers
 */
export default function makeRootReducer(appReducers={}) {
	Object.keys(appReducers).forEach(reducer => {
		if (reducer in platformReducers) {
			throw new Error(`'${reducer}' is a reserved platform reducer name`);
		}
	});
	return combineReducers({
		...platformReducers,
		...appReducers,
	});
}

