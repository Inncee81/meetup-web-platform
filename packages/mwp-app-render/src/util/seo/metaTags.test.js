import locales from 'mwp-config/locales';
import { MOCK_EVENT } from 'meetup-web-mocks/lib/api';

import {
	generateCanonicalUrl,
} from './links';

import {
	generateMetaData,
	generateGeoMetaData,
} from './metaTags';

const MOCK_META = {
	title: 'mock title',
	description: 'mock description',
	keywords: 'mock keywords',
	baseUrl: 'https://mock-base-url.com',
	route: '/mock/preview/path',
	appPath: '/mock/app/path',
	imageUrl: 'https://www.meetup.com/mock-default-image.jpg',
	localeCode: 'fr-FR',
};

describe('generateMetaData', () => {
	const metaData = generateMetaData(MOCK_META);

	it('should set the title', () => {
		const titleMeta = metaData.filter(
			obj => obj.content === MOCK_META.title && obj.property === 'og:title'
		);
		expect(titleMeta).toHaveLength(1);
	});

	it('should set the description', () => {
		const descriptionMeta = metaData.filter(
			obj =>
				obj.content === MOCK_META.description &&
				(obj.property === 'og:description' || obj.name === 'description')
		);
		expect(descriptionMeta).toHaveLength(2);
	});

	it('should set the keywords', () => {
		const keywordsMeta = metaData.filter(
			obj => obj.content === MOCK_META.keywords && obj.name === 'keywords'
		);
		expect(keywordsMeta).toHaveLength(1);
	});

	it('should set the app url', () => {
		const appUrlMeta = metaData.filter(obj => obj.property === 'al:web:url');
		expect(appUrlMeta).toHaveLength(1);
	});

	it('should set the appPath', () => {
		const appPathMeta = metaData.filter(obj => obj.content === MOCK_META.appPath);
		expect(appPathMeta).toHaveLength(2);
	});

	it('should set the url as localized url', () => {
		const expectedUrl = generateCanonicalUrl(
			MOCK_META.baseUrl,
			MOCK_META.localeCode,
			MOCK_META.route
		);
		const ogUrlMeta = metaData.filter(
			obj => obj.content === expectedUrl && obj.property === 'og:url'
		);
		const alUrlMeta = metaData.filter(
			obj => obj.content === expectedUrl && obj.property === 'al:web:url'
		);
		expect(ogUrlMeta).toHaveLength(1);
		expect(alUrlMeta).toHaveLength(1);
	});

	describe('open graph meta data', () => {
		it('should set og:site_name to "Meetup"', () => {
			const ogSiteName = obj =>
				obj.content === 'Meetup' && obj.property === 'og:site_name';
			expect(ogSiteName).toHaveLength(1);
		});
		it('should set og:type to "article"', () => {
			const metaData = generateMetaData(MOCK_META);
			const ogType = metaData.filter(
				obj => obj.content === 'article' && obj.property === 'og:type'
			);
			expect(ogType).toHaveLength(1);
		});
		it('should set og:url', () => {
			const urlMeta = metaData.filter(obj => obj.property === 'og:url');
			expect(urlMeta).toHaveLength(1);
		});
		it('should set og:image', () => {
			const ogImage = metaData.filter(
				obj => obj.content === MOCK_META.imageUrl && obj.property === 'og:image'
			);
			expect(ogImage).toHaveLength(1);
		});
	});
	describe('twitter meta data', () => {
		it('should set twitter:image', () => {
			const twitterImage = metaData.filter(
				obj =>
					obj.content === MOCK_META.imageUrl && obj.property === 'twitter:image'
			);
			expect(twitterImage).toHaveLength(1);
		});
	});
});

describe('generateGeoMetaData', () => {
	it('should set position', () => {
		const lat = 'hello world';
		const lon = 'hello world';
		const processedObj = generateGeoMetaData({ lat, lon });
		const filteredObj = processedObj.filter(
			obj => obj.content === `${lat};${lon}` && obj.property === 'geo.position'
		);
		expect(filteredObj).toHaveLength(1);
	});

	it('should NOT geo position', () => {
		const processedObj = generateGeoMetaData({});
		const filteredObj = processedObj.filter(obj => obj.property === 'geo.position');
		expect(filteredObj).toHaveLength(0);
	});

	it('should set place w/ city', () => {
		const city = 'hello world';
		const processedObj = generateGeoMetaData({ city });
		const filteredObj = processedObj.filter(
			obj => obj.content === city && obj.property === 'geo.placename'
		);
		expect(filteredObj).toHaveLength(1);
	});

	it('should set place w/ city + state', () => {
		const city = 'hello world';
		const state = 'hello world';
		const processedObj = generateGeoMetaData({ city, state });
		const filteredObj = processedObj.filter(
			obj => obj.content === `${city}, ${state}` && obj.property === 'geo.placename'
		);
		expect(filteredObj).toHaveLength(1);
	});

	it('should set place w/ city + state + country', () => {
		const city = 'hello world';
		const state = 'hello world';
		const country = 'USA';
		const processedObj = generateGeoMetaData({ city, state, country });
		const filteredObj = processedObj.filter(
			obj =>
				obj.content === `${city}, ${state}, ${country}` &&
				obj.property === 'geo.placename'
		);
		expect(filteredObj).toHaveLength(1);
	});

	it('should set place w/ city + state + country', () => {
		const country = 'USA';
		const processedObj = generateGeoMetaData({ country });
		const filteredObj = processedObj.filter(
			obj => obj.content === country && obj.property === 'geo.region'
		);
		expect(filteredObj).toHaveLength(1);
	});
});
