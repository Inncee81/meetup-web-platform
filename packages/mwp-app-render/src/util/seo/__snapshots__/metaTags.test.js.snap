// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateGeoMetaData matches snap 1`] = `
Array [
  Object {
    "content": "mock-lat;mock-lon",
    "property": "geo.position",
  },
  Object {
    "content": "mock-city, mock-state, mock-country",
    "property": "geo.placename",
  },
  Object {
    "content": "mock-country",
    "property": "geo.region",
  },
]
`;

exports[`generateGeoMetaData should *not* set geo.position when lat or lon are missing 1`] = `
Array [
  Object {
    "content": "mock-city, mock-state, mock-country",
    "property": "geo.placename",
  },
  Object {
    "content": "mock-country",
    "property": "geo.region",
  },
]
`;

exports[`generateGeoMetaData should exclude state and country from geo.placename when not provided 1`] = `
Array [
  Object {
    "content": "mock-lat;mock-lon",
    "property": "geo.position",
  },
  Object {
    "content": "mock-city",
    "property": "geo.placename",
  },
  Object {
    "content": undefined,
    "property": "geo.region",
  },
]
`;

exports[`generateGeoMetaData should exclude state from geo.placename when not provided 1`] = `
Array [
  Object {
    "content": "mock-lat;mock-lon",
    "property": "geo.position",
  },
  Object {
    "content": "mock-city, mock-country",
    "property": "geo.placename",
  },
  Object {
    "content": "mock-country",
    "property": "geo.region",
  },
]
`;

exports[`generateMetaData matches snap 1`] = `
Array [
  Object {
    "content": "mock description",
    "name": "description",
  },
  Object {
    "content": "mock,keywords,comma,separated",
    "name": "keywords",
  },
  Object {
    "content": "36212898404",
    "property": "fb:app_id",
  },
  Object {
    "content": "Meetup",
    "property": "og:site_name",
  },
  Object {
    "content": "article",
    "property": "og:type",
  },
  Object {
    "content": "mock title",
    "property": "og:title",
  },
  Object {
    "content": "mock description",
    "property": "og:description",
  },
  Object {
    "content": "https://mock-base-url.com/fr-FR/mock/preview/path",
    "property": "og:url",
  },
  Object {
    "content": "https://www.meetup.com/mock-default-image.jpg",
    "property": "og:image",
  },
  Object {
    "content": "Meetup",
    "property": "al:android:app_name",
  },
  Object {
    "content": "com.meetup",
    "property": "al:android:package",
  },
  Object {
    "content": "375990038",
    "property": "al:ios:app_store_id",
  },
  Object {
    "content": "Meetup",
    "property": "al:ios:app_name",
  },
  Object {
    "content": "true",
    "property": "al:web:should_fallback",
  },
  Object {
    "content": "/mock/app/path",
    "property": "al:android:url",
  },
  Object {
    "content": "/mock/app/path",
    "property": "al:ios:url",
  },
  Object {
    "content": "https://mock-base-url.com/fr-FR/mock/preview/path",
    "property": "al:web:url",
  },
  Object {
    "content": "summary_large_image",
    "property": "twitter:card",
  },
  Object {
    "content": "@meetup",
    "property": "twitter:site",
  },
  Object {
    "content": "https://www.meetup.com/mock-default-image.jpg",
    "property": "twitter:image",
  },
  Object {
    "content": "mock description",
    "property": "twitter:description",
  },
  Object {
    "content": "375990038",
    "property": "twitter:app:id:iphone",
  },
  Object {
    "content": "Meetup",
    "property": "twitter:app:name:iphone",
  },
  Object {
    "content": "375990038",
    "property": "twitter:app:id:ipad",
  },
  Object {
    "content": "Meetup",
    "property": "twitter:app:name:ipad",
  },
  Object {
    "content": "com.meetup",
    "property": "twitter:app:id:googleplay",
  },
  Object {
    "content": "Meetup",
    "property": "twitter:app:name:googleplay",
  },
]
`;
